### PR TITLE
Creating a singleton instance of Contentful

### PR DESCRIPTION
**Pull request by Bounteous17**
## Description

- Being able to instance a client and retrieve content from Contentful
- Next PR would implement a generic output format after parsing CMS retrieved data no matter which provider is (Contentful, Strapim etc ...)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by matyasjay — COMMENTED** [review]
Looks good to me so far! One thing I would add is the type infers from generics instead explicit `any`. 

I'm thinking of something like:

```ts
 public async getEntry<T = Record<string, unknown>>(entryId: string) {
    return await this.getEntryHarmonized<T>(
      () => this.getClientInstance().getEntry<T>(entryId),
      // The following might not even needed if we can infer `T` from the return type
      ({ fields }: { fields: T }) => ({ data: fields }),
    );
  }
```

This requires typed client instances. The above example is a high level abstraction of what actually needed here. For now, I'd say we can merge this as is.

File: __tests__/e2e/cms/contentful.test.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { config as configDotenv } from 'dotenv';
+configDotenv();
+
+import { Contentful } from '../../../src/cms/contentful';
+
+describe('contentful', () => {
+  describe('should not fail', () => {
```
Comment (inigomarquinez) [inline __tests__/e2e/cms/contentful.test.ts:7]: Should we add a test for a failing case? e.g. trying to fetch a non-existing entry

**Review by inigomarquinez — APPROVED** [review]
I really like your approach, @Bounteous17 

I've just added a small suggestion about the tests, but we can do it in a future PR if you want.

File: __tests__/e2e/cms/contentful.test.ts (hunk 1 above)
Comment (matyasjay) [reply __tests__/e2e/cms/contentful.test.ts:7]: I assume the CF API has a well declared return type for such case. Although, it is defo a good idea to check if we also return the expected value for that.

**Comment by inigomarquinez** [comment]
@UlisesGascon , perhaps your feedback is useful here related to the alerts reported by socket.dev related to some dependencies running post-install scripts.

**Review by UlisesGascon — APPROVED** [review]
I only reviewed the dependencies and I left a comment with my analysis.

**Comment by UlisesGascon** [comment]
@SocketSecurity ignore npm/spawn-sync@1.0.15
@SocketSecurity ignore npm/contentful-cli@3.1.43

**Review by UlisesGascon — CHANGES_REQUESTED** [review]
Not sure why, but I just realized that the `package-lock.json` is different on my machine.

Steps to reproduce:
- (Optional) delete the `node_modules`
- `nvm use`. In my case node v20.11.0 (npm v10.2.4)
- `npm i`

And here is the discrepancy:
Screenshot 2024-02-14 at 14 11 26

**Comment by inigomarquinez** [comment]
> ### Conclusion
> Overall, both postscripts does not seem malicious or with known vulnerabilities (AFIK), but I will suggest to install them using `--ignore-scripts` flag in npm. But as this is a library that will be distributed, I am not sure if the end users will take care of this or not.

Thank you so much @UlisesGascon !

**Comment by matyasjay** [comment]
> ### Conclusion
> Overall, both postscripts does not seem malicious or with known vulnerabilities (AFIK), but I will suggest to install them using `--ignore-scripts` flag in npm. But as this is a library that will be distributed, I am not sure if the end users will take care of this or not.

Hi @UlisesGascon, thanks for your time and valuable info! One thing here: `--ignore-scripts` might be not optimal but as I mentioned to @inigomarquinez internally, running CI with `--no-optional` flag would be sufficient to ignore the dependency since `contentful-cli` defined as optional.

**Comment by matyasjay** [comment]
> Not sure why, but I just realized that the `package-lock.json` is different on my machine.
> 
> Steps to reproduce:
> 
> * (Optional) delete the `node_modules`
> * `nvm use`. In my case node v20.11.0 (npm v10.2.4)
> * `npm i`
> 
> And here is the discrepancy: Screenshot 2024-02-14 at 14 11 26

This is the result of you are running raw `npm i`. However in CI, optional deps should be ignored hence no `contentful-cli` in the lock file. You can repro by running the install with `--no-optional` flag.

**Comment by Bounteous17** [comment]
> changes

@matyasjay @inigomarquinez  we would implement the abstract type for the different providers instance on another PR. For the moment let's leave it as an `any`

**Comment by Bounteous17** [comment]
> Looks good to me so far! One thing I would add is the type infers from generics instead explicit `any`.
> 
> I'm thinking of something like:
> 
> ```ts
>  public async getEntry<T = Record<string, unknown>>(entryId: string) {
>     return await this.getEntryHarmonized<T>(
>       () => this.getClientInstance().getEntry<T>(entryId),
>       // The following might not even needed if we can infer `T` from the return type
>       ({ fields }: { fields: T }) => ({ data: fields }),
>     );
>   }
> ```
> 
> This requires typed client instances. The above example is a high level abstraction of what actually needed here. For now, I'd say we can merge this as is.

@matyasjay @inigomarquinez  we would implement the abstract type for the different providers instance on another PR. For the moment let's leave it as an `any`

**Comment by Bounteous17** [comment]
> how

@UlisesGascon At the beginning `contentful-cli` was part of the `dependencies` until I realized that it should be part of `optionalDependencies`, that's the fix: https://github.com/onebeyond/agnostic-cms-harmonizer/pull/5/commits/1cf98204789803c5ab30122345925d6db8b4c900

**Comment by Bounteous17** [comment]
> Not sure why, but I just realized that the `package-lock.json` is different on my machine.
> 
> Steps to reproduce:
> 
> * (Optional) delete the `node_modules`
> * `nvm use`. In my case node v20.11.0 (npm v10.2.4)
> * `npm i`
> 
> And here is the discrepancy: Screenshot 2024-02-14 at 14 11 26

@UlisesGascon At the beginning `contentful-cli` was part of the `dependencies` until I realized that it should be part of `optionalDependencies`, that's the fix: https://github.com/onebeyond/agnostic-cms-harmonizer/pull/5/commits/1cf98204789803c5ab30122345925d6db8b4c900

**Review by inigomarquinez — APPROVED** [review]

**Review by matyasjay — APPROVED** [review]

**Comment by inigomarquinez** [comment]
@all-contributors please add @Bounteous17 for code,test,doc
@all-contributors please add @matyasjay for review
@all-contributors please add @UlisesGascon for review

**Comment by inigomarquinez** [comment]
@all-contributors please add @UlisesGascon for review

**Comment by inigomarquinez** [comment]
https://github.com/all-contributors please add @UlisesGascon for review

**Comment by inigomarquinez** [comment]
@all-contributors please add @UlisesGascon for review

**Pull request merged**